### PR TITLE
Add support for `httpd`

### PR DESCRIPTION
**Pull request by fadushin**
This PR adds support for the httpd family of modules, allowing applications to start a simple HTTP daemon with the following features:
* Ability to write callbacks for REST-based APIs
* Ability to serve files from the `priv` section of an AVM file
* Ability to communicate with web clients over the web sockets protocol

### CODE REVIEW

File: examples/gen_tcp_server_example/rebar.config [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{erl_opts, [debug_info]}.
+{deps, [
+    {atomvm_lib, {git, "https://github.com/fadushin/atomvm_lib.git", {branch, "master"}}}
```
Comment (UncleGrumpy) [inline examples/gen_tcp_server_example/rebar.config]: URL needs update to github.com/atomvm/atomvm_lib

File: examples/httpd_example/priv/index.html [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+for which I would be most grateful, but you are not required to do so.
+If we meet some day, and you think this stuff is worth it, you can buy me a
+beer in return.   Fred Dushin
+----------------------------------------------------------------------------
+-->
+<!DOCTYPE html>
+<html>
+	<meta charset="UTF-8">
+	<meta name="viewport" content="width=device-width, initial-scale=1">
+	<!-- <link rel="stylesheet" href="https://www.w3schools.com/w3css/4/w3.css"> -->
```
Comment (UncleGrumpy) [inline examples/httpd_example/priv/index.html]: Is this supposed to be commented out?

File: examples/httpd_example/priv/index.html [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+					<button class="w3-bar-item w3-button" onclick="openPanel(event, 'memory')">Memory</button>
+					<div id="icon"><img src="favicon.ico" alt="AtomVM logo" width="30" height="30" /></div>
+				</div>
+
+				<div id="about" class="w3-panel w3-animate-opacity wc-panel">
+					<div id="about-view">
+						Welcome to the AtomVM web console.
+
+						Use this console to explore the current state of your device.
+						<!-- <img src="AVM-logo.png" alt="AtomVM logo" width="300" height="300"/> -->
```
Comment (UncleGrumpy) [inline examples/httpd_example/priv/index.html]: Was this something that should be removed? ...or uncommented?

File: examples/httpd_example/priv/index.html [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+						</script>
+					</div>
+				</div>
+			</div>
+			<div id="footer" class="w3-row w3-teal">
+				<div class="w3-container w3-threequarter">
+					AtomVM Web Console
+				</div>
+				<div class="w3-container w3-quarter">
+					<!-- <strong>version:</strong> <em>master</em> -->
```
Comment (UncleGrumpy) [inline examples/httpd_example/priv/index.html]: same question.

File: examples/httpd_example/priv/index.html [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+			}
+			function timedShowContents() {
+				setTimeout(showContents, 1000);
+			}
+			function showContents() {
+				document.getElementById("loader").style.display = "none";
+				document.getElementById("content").style.display = "block";
+			}
+			function openApConfigModal(){
+				/*
```
Comment (UncleGrumpy) [inline examples/httpd_example/priv/index.html]: Should this also be removed, or maybe a comment about this does if uncommented.

File: examples/httpd_example/priv/js/app.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+	parse: function(data) {
+		data["chip_model"] = data.esp32_chip_info["model"];
+		data["chip_cores"] = data.esp32_chip_info["cores"];
+		data["chip_features"] = data.esp32_chip_info["features"];
+		data["chip_revision"] = data.esp32_chip_info["revision"];
+		return data;
+	},
+
+	// convert_frequency: function(freq) {
```
Comment (UncleGrumpy) [inline examples/httpd_example/priv/js/app.js]: more old code?

File: examples/httpd_example/priv/js/app.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+		var html = this.template(this.model.toJSON());
+		this.$el.html(html);
+		return this;
+	}
+});
+var system_view = new SystemView({model: system_model})
+
+
+
+// var precision2 = function(number){
```
Comment (UncleGrumpy) [inline examples/httpd_example/priv/js/app.js]: same question.

File: src/httpd.erl [hunk 8]
```diff
@@ -17,144 +17,454 @@
...
+            io:format("Caught error: ~p:~p:~p~n", [A, E, S]),
+            {close, create_error(?BAD_REQUEST, E)}
+    end.
+
+%% @private
+handle_http_request(Socket, Packet, State) ->
+    case maps:get(Socket, State#state.pending_request_map, undefined) of
+        undefined ->
+            HttpRequest = parse_http_request(binary_to_list(Packet)),
+            % ?TRACE("HttpRequest: ~p~n", [HttpRequest]),
```
Comment (UncleGrumpy) [inline src/httpd.erl:128]: Commented out intentionally?

**Review by UncleGrumpy — COMMENTED** [review]
Aside from one outdated URL I didn't spot any problems, I pointed out some commented out code that may or may not have been left in intentionally.

File: examples/gen_tcp_server_example/rebar.config (hunk 1 above)
Comment (fadushin) [reply examples/gen_tcp_server_example/rebar.config]: Fixed and in the httpd example, which was also wrong (not in hex!)

File: examples/httpd_example/priv/index.html (hunk 3 above)
Comment (fadushin) [reply examples/httpd_example/priv/index.html]: Removed all that commented out code.  Thanks for spotting.

**Review by UncleGrumpy — APPROVED** [review]
Looks Good.

**Pull request merged**